design/one-pager-managed-resource-api-design.md

```diff
@@ -0,0 +1,294 @@
+
+In most of the cases, an application setup requires the cloud resources to be connected/related in a custom way for
+various configurations. An example of this could be resource grouping which could happen as a physical layer
+(VPC network) or logical (resource grouping in providers). So, one can assume that in most of the cases where you have
+more than one service in your setup, you want to relate one to another, hence you need to _refer_ to one another in
+their configurations. There could be one or many configuration fields in any resource that we may not know beforehand.
+
+The easiest answer to solve this is probably just exposing everything the provider exposes. This is a good starting
+point but not a complete solution. A given resource can be either _referrer_ or _referred_ in a setup. In the former,
+we're talking about how we can refer to another resource in the `Spec` but for the latter, it's about how best this
+resource can be referred in another resource's `Spec`. There is also abundance of fields that are not necessary for either
+scenario but are necessary on a level that is higher than Crossplane, like creation timestamps on a Web UI built on top
+of Crossplane.
+
+Ideally, what we want is to expose all possible configurations in `Spec` and whatever might be needed by the output 
+users of this resource in the `Status`. Generally, it's better to just copy all fields from the provider's API as
+starting point and eliminate them one by one if needed.
+
+What goes into `Spec`:
+Does the provider allow configuration of this parameter at _any_ point in the life cycle of the resource? Include if yes.
+This includes the fields that are late-initialized, meaning some fields could be populated by the provider when you
+actually create the resource but it also gives you the ability to change them. An example for this could be
+auto-generated resource tags. If the provider tags the resource without us telling them to do so, controller should
+update `Spec` and input user should make changes on that current value of the field. Related to https://github.com/crossplaneio/crossplane-runtime/issues/10
+
+What goes into `Status`:
+Does the provider give the information of that field on the first level of the query, `GET`?
+For example, `network` could be a top level information in `GET` request of a database resource but the details of that
+network should not be included via making extra calls, we want only the information that is present on the top level.
+If there is a reference to another provider API object (the object could be a tree of other objects like in Azure), only
+include the provider's identifier of that object to make it easier to query it when needed. Trying to include the whole
+tree can even go to the state where all your setup that is connected to each other is represented under one resource's
+status, which is a very bad UX.
```

Comment:
My understanding is that we may need to make a recommendation around this Azure-ish case for spec inputs as well as status outputs. Two examples of this are:

* Creating or updating an [Azure virtual network](https://docs.microsoft.com/en-us/rest/api/virtualnetwork/virtualnetworks/createorupdate). The Azure API allows you to specify a set of subnets and peerings that also exist as distinct Azure API objects and thus _could_ be modeled as distinct managed resources.
* Creating or updating a [GKE cluster](https://cloud.google.com/kubernetes-engine/docs/reference/rest/v1/projects.locations.clusters/create). The API allows you to provide a set of node pools, that also exist as distinct GKE API objects and thus _could_ be modeled as distinct managed resources.

Perhaps Crossplane should take a stance something like "if this external resource could be CRUD as either a subfield of a parent resource _or_ as a distinct resource, Crossplane supports only the latter". So basically we'd omit the `nodePools` of a GKE cluster managed resource and instead require you to create distinct node pool managed resources instead.

Reply:
> if this external resource could be CRUD as either a subfield of a parent resource or as a distinct resource, Crossplane supports only the latter

I thought about this a lot and I think we might _have to_ do this way even though we want to support referring resources that are not managed by the given Crossplane instance or Crossplane has no support for that resource type yet.

Let's say we created a struct that has identification information of []*SubnetReference and included this as a field in Azure VirtualNetwork to support adding non-Crossplane-managed Subnet resource.

```
type VirtualNetwork struct {
  subnets []*SubnetReference
}
```
I created a VirtualNetwork object with this array having a subnet named `A`. Azure will create the VirtualNetwork and the Subnet `A`. Then I went ahead and created an actual Subnet CR with subnet name `B` that refers to my VirtualNetwork. In that case my desired state about how VirtualNetwork's subnets should be constructed is divided; virtual network controller will delete `B` since it sees the desired state of VirtualNetwork having only `A` subnet, on the other hand, subnet controller will say `B` who refers to my VirtualNetwork should exist. So, possibly, one will delete and the other one will create.

In the current implementation, VirtualNetwork object doesn't mention any Subnet at all. So, the desired state of Subnet part of the VirtualNetwork is managed only by our Subnet resource and that is consistent. So, I guess we'll either expand VirtualNetwork resource to make it the only way of creating Subnets or we will eliminate other resource references in it.

We can expand VirtualNetwork and remove Subnet, but we don't know where this might end. It can be a really long tail of resources provisioned in only one CR. So, we might have to take that stance where we say you have to create a CR for that sub-resource if we have support for it. I think having external resource policy read-only mode(https://github.com/crossplaneio/crossplane-runtime/issues/22#issuecomment-526459345) will help users to add resource only for reference purposes even though they don't want to manage them with Crossplane or with _that_ Crossplane instance.

I think this analogy applies to GKE node pools as well.

Reply:
One thing that came up when syncing with @muvaf on this topic was that, as mentioned, we are currently ignoring the embedded `Subnet` slice on the Azure `VirtualNetwork`, so adding a subnet via a `Subnet` custom resource does not cause the `VirtualNetwork` controller to update the `VirtualNetwork`. However, I am concerned that if some other part of the `VirtualNetwork` changes that causes the controller to update, that it may delete any subnets that were created. The subnets would then be recreated by the `Subnet` controller, but it could have some pretty severe negative effects on other resources. It is possible that the Azure API will not delete subnets if update is called on a virtual network, but I am not certain. I am going to test this out today.

Reply:
@muvaf @negz I confirmed this behavior and have opened an issue to track crossplaneio/stack-azure#23